### ISSUE TITLE
Template matching exception

### ISSUE THREAD

**Issue by strarsis** [peterbraden/node-opencv#417]
Using matchTemplate(...) results in an exception, 
albeit template image is smaller than original image, 
images were grayscaled and 
colour depth of both images being equivalent:
````
OpenCV Error: Assertion failed (corrsize.height <= img.rows + templ.rows - 1 && corrsize.width <= img.cols + templ.cols - 1) in crossCorr, file /home/build/src/CV/opencv/OpenCV/modules/imgproc/src/templmatch.cpp, line 658
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/build/src/CV/opencv/OpenCV/modules/imgproc/src/templmatch.cpp:658: error: (-215) corrsize.height <= img.rows + templ.rows - 1 && corrsize.width <= img.cols + templ.cols - 1 in function crossCorr

Aborted
````

Relevant code with example images used can be found here for reproducing this issue:
[https://github.com/strarsis/node-opencv-templatematching-test](https://github.com/strarsis/node-opencv-templatematching-test)

**Comment by pkallos** [comment]
that's the kind of message I would expect to see if your original image was smaller than the template, but looking at your github code it looks correct....

FWIW I am using this template matching to good effect, albeit with opencv3+ and node-opencv pegged to `"opencv": "^5.0.0"`

Also worth noting that open cv3 adds support for RGB so you don't need to grayscale those anymore, give it a shot

**Comment by strarsis** [comment]
It appears that by default OpenCV 2.x is used by the opencv package, 
with opencv 5.0.0 I get the error message.
Do I have to rebuild the opencv package against OpenCV 3.x?

**Comment by pkallos** [comment]
I would recommend it yeah, the RGB channels in the templateMatching in OpenCV3 give much better results anyways.

**Comment by strarsis** [comment]
I got libopencv-dev 2.6 on Debian Jessie installed.
Are there instructions/README on how to (re)build node-opencv 
against OpenCV 3.x on Linux (Debian Jessie)?

**Comment by pkallos** [comment]
Here is a redacted section of our Dockerfile (sorry if I introduced a syntax error during the redaction, but you'll get the idea):

```dockerfile
FROM ubuntu:14.04

# Install OpenCV 3.0 and nodejs in one go
RUN apt-get -y update && \
   export BUILD_PACKAGES="git wget curl build-essential cmake" && \
   export DEPS="libgtk2.0-dev pkg-config libavcodec-dev libavformat-dev libswscale-dev libtbb2 libtbb-dev libjpeg-dev libpng-dev libtiff-dev libjasper-dev libdc1394-22-dev ocl-icd-opencl-dev libcanberra-gtk3-module libeigen3-dev" && \
  apt-get install -y $BUILD_PACKAGES $DEPS && \
  cd / && \
  git clone --depth 1 https://github.com/Itseez/opencv.git && \
  cd /opencv && \
  git checkout 3.0.0 && \
  mkdir /opencv/build && \
  cd /opencv/build && \
  cmake -D CMAKE_BUILD_TYPE=RELEASE \
      -D CMAKE_INSTALL_PREFIX=/usr/local \
      -D WITH_OPENGL=ON \
      -D WITH_TBB=OFF \
      -D BUILD_EXAMPLES=ON \
      -D BUILD_NEW_PYTHON_SUPPORT=ON \
      -D WITH_V4L=ON \
      .. && \
  make -j$(nproc) && \
  make install && \
  rm -rf /opencv && \
  cd /tmp && \
  wget http://nodejs.org/dist/node-latest.tar.gz && \
  tar xvzf node-latest.tar.gz && \
  rm -f node-latest.tar.gz && \
  cd node-v* && \
  ./configure && \
  CXX="g++ -Wno-unused-local-typedefs" make && \
  CXX="g++ -Wno-unused-local-typedefs" make install && \
  cd /tmp && \
  rm -rf /tmp/node-v* && \
  npm install -g npm && \
  echo -e '\n# Node.js\nexport PATH="node_modules/.bin:$PATH"' >> /root/.bashrc && \
  apt-get install -y tesseract-ocr && \
  apt-get -y autoremove && \
  apt-get remove --purge -y $BUILD_PACKAGES

# Install app dependencies
COPY package.json /usr/src/app/
RUN npm install
```

TLDR; build everything from source so you can avoid `apt` repository headaches.

**Comment by strarsis** [comment]
The git clone command seems to cause trouble - tag '3.0.0' cannot be found/checked out 
and there are no tags at all in the local repository:
````
[...]
git clone --depth 1 https://github.com/Itseez/opencv.git #&& \
[...]
````

**Comment by pkallos** [comment]
fetch the tags then

**Comment by strarsis** [comment]
OpenCV 3.x has been successfully built + installed (binaries and libraries are available).
Can I just use `npm install opencv` in project directory or do I have to build+install node-opencv also manually?

**Comment by pkallos** [comment]
I think from here you should be able to `npm install opencv` and it will build referencing your new openCV3 libs! Let me know if it works.

**Comment by strarsis** [comment]
First I removed the opencv lib packages from Debian repository, to be sure that nothing old is used.
Then I removed the node_modules folder/ of the project and installed node-opencv again.
The opencv binaries from the opencv 3.x build use the opencv 3.x according to ldd:
````
        libopencv_core.so.3.0 => /usr/local/lib/libopencv_core.so.3.0 (0x00007fad7699c000)
        libopencv_highgui.so.3.0 => /usr/local/lib/libopencv_highgui.so.3.0 (0x00007fad76790000)
        libopencv_imgproc.so.3.0 => /usr/local/lib/libopencv_imgproc.so.3.0 (0x00007fad75801000)
        libopencv_imgcodecs.so.3.0 => /usr/local/lib/libopencv_imgcodecs.so.3.0 (0x00007fad755c9000)
        libopencv_videoio.so.3.0 => /usr/local/lib/libopencv_videoio.so.3.0 (0x00007fad7539b000)
````

**Comment by pkallos** [comment]
Hopefully this unblocks you: strarsis/node-opencv-templatematching-test#1

**Comment by strarsis** [comment]
Awesome! It works now.

**Comment by strarsis** [comment]
So the most important modification was to pass the path to an image - not a buffer?
What if I want to use a buffer instead?

**Comment by pkallos** [comment]
I have no idea 😄  

[Looking at the code a bit, it seems like it only accepts a filename](https://github.com/peterbraden/node-opencv/blob/13a1de4f9259b5e745021f3198a8339ac99edecd/src/Matrix.cc#L2246-L2254) ... I don't think the API supports a buffer at all.
